### PR TITLE
SonarQube on Maven and Gradle can fail the build on quality gate fail…

### PR DESCRIPTION
**Pull request by HSAR**
…ure.

### CODE REVIEW

File: Tests/L0/Maven/_suite.ts [hunk 1]
```diff
@@ -121,6 +121,24 @@ function createTempDirsForSonarQubeTests(): void {
     }
 }
 
+function captureStream(stream):{unhook():void, captured():string} {
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: This function handles the business of capturing _process.stdout_ for the duration of a test. It redirects all calls to the _write()_ function of the given stream until _unhook()_ is called. All redirected calls are stored in the internal buffer, which is exposed and can be asserted on.

File: Tests/L0/Gradle/_suite.ts [hunk 2]
```diff
@@ -1163,10 +1163,8 @@ describe('gradle Suite', function () {
                     'gradlew build sonarqube -I /Gradle/CodeAnalysis/sonar.gradle -Dsonar.host.url=http://sonarqube/end/point -Dsonar.login=uname -Dsonar.password=pword -Dsonar.projectName=test_sqProjectName -Dsonar.projectKey=test_sqProjectKey -Dsonar.projectVersion=test_sqProjectVersion -Dsonar.analysis.mode=issues -Dsonar.report.export.path=sonar-report.json'
                 ), 'should have run the gradle wrapper with the appropriate SonarQube arguments');
 
-                assert(tr.stdout.indexOf('task.addattachment type=Distributedtask.Core.Summary;name=SonarQube Analysis Report') > -1,
-                    'should have uploaded a SonarQube Analysis Report build summary');
-                assertSonarQubeBuildSummaryContains(testStgDir,
-                    'Detailed SonarQube reports are not available for pull request builds.');
```
Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: Change of behaviour: on PR builds no build summary is uploaded at all.

File: Tests/L0/Maven/_suite.ts [hunk 3]
```diff
@@ -153,6 +176,7 @@ describe('Maven Suite', function () {
     this.timeout(20000);
 
     before((done) => {
+        Q.longStackSupport = true;
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: A debug option that has memory and performance impacts but improves stack trace behaviour.

File: Tests/L0/Maven/_suite.ts [hunk 4]
```diff
@@ -985,10 +1008,8 @@ describe('Maven Suite', function () {
                 assert(tr.stderr.length < 1, 'should not have written to stderr');
                 assert(tr.succeeded, 'task should have succeeded');
 
-                assert(tr.stdout.indexOf('task.addattachment type=Distributedtask.Core.Summary;name=SonarQube Analysis Report') > 0,
-                    'should have uploaded a SonarQube Analysis Report build summary');
-                assertSonarQubeBuildSummaryContains(testStgDir,
-                    'Detailed SonarQube reports are not available for pull request builds.');
```
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Change of behaviour: on PR builds no build summary is uploaded at all.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts [hunk 5]
```diff
@@ -21,145 +23,42 @@ export function isSonarQubeAnalysisEnabled(): boolean {
...
+/**
+ * Apply all parameters necessary for SonarQube operation to the given ToolRunner.
+ * @param toolRunner
+ * @returns {ToolRunner}
+ */
+export function applySonarQubeParameters(toolRunner: ToolRunner): ToolRunner {
+    return SonarQubeParameterHelper.applySonarQubeParameters(toolRunner);
 }
 
 // Upload a build summary with links to available SonarQube dashboards for further analysis details.
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Mistakenly left this comment behind when re-writing the docs. Will delete.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts [hunk 6]
```diff
@@ -21,145 +23,42 @@ export function isSonarQubeAnalysisEnabled(): boolean {
...
+    var sqRunSettings:SonarQubeRunSettings = getSonarQubeRunSettings(sqBuildFolder);
+    var sqMetrics:SonarQubeMetrics = getSonarQubeMetrics(sqRunSettings);
+
+    // Wait for all promises to complete before proceeding (even if one or more promises reject).
+    return Q.allSettled([
+        processSonarQubeBuildSummary(sqRunSettings, sqMetrics),
+        processSonarQubeBuildBreaker(sqRunSettings, sqMetrics),
+    ])
+        .then(() => {
+        })
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Previously, the only integration task was the upload the build summary. Now that we have more, the public entry point will be this function _processSonarQubeIntegration()_, which will act on all enabled integration actions.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts [hunk 7]
```diff
@@ -172,52 +71,108 @@ function isPrBuild(): boolean {
...
+/**
+ * Returns, as an object, the contents of the report-task.txt file created by SonarQube plugins
+ * The returned object contains the following properties:
+ *     projectKey, serverUrl, dashboardUrl, ceTaskId, ceTaskUrl
+ * @param sonarPluginFolder The folder created by SonarQube integration during build
+ * @returns {SonarQubeRunSettings} An object with fields corresponding to the properties exposed in report-task.txt, or null if PR build
+ */
+function getSonarQubeRunSettings(sonarPluginFolder: string): SonarQubeRunSettings {
+    if (isPrBuild()) {
+        return null;
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Returning null is a deliberate decision here. The report-task.txt file does not exist in PR builds (it is the only scenario where we expect to not find it), but there may still be some integration actions that we can do without it.

**Comment by HSAR** [comment]
As we are currently manually synchronising the Maven/CodeAnalysis/SonarQube and Gradle/CodeAnalysis/SonarQube folders, please note that all code changes in those folders will be duplicated.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts [hunk 8]
```diff
@@ -1,5 +1,64 @@
...
+/// <reference path="../../../../definitions/vsts-task-lib.d.ts" />
+
+import tl = require('vsts-task-lib/task');
+
 export class SonarQubeEndpoint {
     constructor(public Url: string, public Username: string, public Password: string) {
     }
+
+    public static createSonarQubeEndpoint():SonarQubeEndpoint {
+        var errorMessage = "Could not decode the generic endpoint. Please ensure you are running the latest agent (min version 0.3.2)";
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: I don't think this is needed anymore. Also, it isn't localized. I'd remove it and assert instead.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts [hunk 9]
```diff
@@ -1,5 +1,64 @@
-// Simple data class for a SonarQube generic endpoint
+/// <reference path="../../../../definitions/vsts-task-lib.d.ts" />
+
+import tl = require('vsts-task-lib/task');
+
 export class SonarQubeEndpoint {
     constructor(public Url: string, public Username: string, public Password: string) {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: I'd make the ctor private unless you need it for tests.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts [hunk 10]
```diff
@@ -1,5 +1,64 @@
...
     }
+
+    public static createSonarQubeEndpoint():SonarQubeEndpoint {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: I'd rename this to "getTaskSonarQubeEndpoint"

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts [hunk 11]
```diff
@@ -6,13 +6,15 @@ import fs = require('fs');
 
 import {ToolRunner} from 'vsts-task-lib/toolrunner';
 
+import tl = require('vsts-task-lib/task');
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: I think we should take the oportunity and get rid of common.ts entirely and replace it with a SonarQube orchestrator. It should probably have only 2 public methods - one for configuring the tool runner and one for doing all of the post processing work (report, build breaker)

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Resolved.

Removed all checks except for _if (tl.getEndpointUrl == null) ..._, which throws an empty error and makes a debug log.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Resolved, done.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts (hunk 10 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Resolved, done.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts (hunk 11 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: In this proposal, where are the functions _isPrBuild()_, _getStagingDir()_ and _getRunSettings()_?

Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Those are related to the interaction with the build system, but I don;'t think it's worth having a separate object to model that. having them as public static is fine as long as they don't complicate testing. For example in a simple class named "TfsBuild"

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts [hunk 12]
```diff
@@ -172,52 +71,108 @@ function isPrBuild(): boolean {
...
+ * @param sqMetrics     SonarQube metrics for the applicable run
+ * @returns {Q.Promise<void>} Promise resolved when action completes.
  */
-function createSonarQubeBuildSummary(sqBuildFolder:string):Q.Promise<string> {
-    // Task report is not created for PR builds - inform the user in the build summary
-    if (isPrBuild()) {
-        tl.debug('Pull request build, will not attempt to fetch detailed SonarQube report');
-        // Looks like: Detailed SonarQube reports are not available for pull request builds.
-        return Q.when(tl.loc('sqAnalysis_BuildSummaryNotAvailableInPrBuild'));
+function processSonarQubeBuildBreaker(sqRunSettings:SonarQubeRunSettings, sqMetrics:SonarQubeMetrics):Q.Promise<void> {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Think about moving the build breaker stuff in a different class.

File: Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
+/// <reference path="../../../../definitions/vsts-task-lib.d.ts" />
+
+import tl = require('vsts-task-lib/task');
+
+import {ToolRunner} from 'vsts-task-lib/toolrunner';
+
+import {SonarQubeEndpoint} from './endpoint';
+import sqCommon = require('./common');
+
+export class SonarQubeParameterHelper {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts]: The file name convention is ussually ParameterHelper (or lowercase if you insist :)).

**Comment by bgavrilMS** [comment]
Great looking tests. LGTM

File: Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts]: Which? C#? In Javascript there are many naming conventions. I am following jQuery's, though of course there are camelCased conventions in addition to PascalCased as you point out.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts (hunk 12 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: The number of classes with a few functions in is starting to become alarming. I believe that, as the implementation of the build breaker necessarily involves use of the task library, that this code belongs in common (where most of the inherently-untestable task-lib calls reside).

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts (hunk 11 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: I don't think those three listed functions are suitably related enough to put in a class together (unless that class is called "common"). _getStagingDir()_ and _getRunSettings()_ have nothing to do with TFS.

File: Tasks/Gradle/CodeAnalysis/SonarQube/common.ts (hunk 12 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Accepted, in progress. Will move all of the build breaker, build summary and isPR stuff into a class related to the VSTS server side.

Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/common.ts]: Resolved, done.

File: Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/parameter-helper.ts]: Remaining with this-file-name-convention.

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Reverted: Typescript does not allow the 'private' access modifier on constructors.

**Pull request merged**